File: bindings/octave/lib/Image.m
classdef Image
    %Image  A PTIFF image: metadata plus tile access (Octave).
    %
    %   Use `Image.open(path)` (read a TIFF/BigTIFF) and `Image.create(path,
    %   width, height, ...)` (write a new one). Reading exposes the structured
    %   camera calibration via `#camera` and the raw tile bytes via
    %   `#read_tile(col, row)`.
    %
    %   Requires the SWIG module (call ptiff() once first).

    properties
        path
        source_   % raw ptiff_source handle (non-empty when open for reading)
        sink_     % raw ptiff_sink handle (non-empty when open for writing)
        camera_
        desc_
    end

    methods (Static = true)
        function img = open(path)
            %OPEN  Open the TIFF/BigTIFF file at path for reading.
            path = char(path);
            [handle, ~] = ptiff_source_open(path);
            if isempty(handle)
                error('ptiff:ImageOpen', 'ptiff_source_open(%s) failed', path);
            end
            img = Image();
            img.path = path;
            img.source_ = handle;
            try
                cam = Camera.from_path(path);
                img.camera_ = cam;
            catch
                img.camera_ = [];
            end
            img = img.load_descriptor_();
        end

        function img = create(path, width, height, varargin)
            %CREATE  Create a new image at path with the given size.
            %
            %   Name-value options:
            %     pixel_type     : a PTIFF_PIXEL_* value (default UInt8 = 0)
            %     tile_width     : tile width in px  (0 = untiled)
            %     tile_height    : tile height in px (0 = untiled)
            %     channel_count  : channels (default 1)
            %     camera         : an optional Camera to persist alongside
            patht = char(path);
            p = inputParser;
            addParameter(p, 'pixel_type', 0);
            addParameter(p, 'tile_width', 0);
            addParameter(p, 'tile_height', 0);
            addParameter(p, 'channel_count', 1);
            addParameter(p, 'camera', []);
            parse(p, varargin{:});

            desc = new_ptiff_image_descriptor();
            desc.width = int32(width);
            desc.height = int32(height);
            desc.pixel_type = int32(p.Results.pixel_type);
            desc.channel_count = int32(p.Results.channel_count);
            if p.Results.tile_width > 0 && p.Results.tile_height > 0
                desc.has_tile_info = int32(1);
                desc.tile_info.tile_width = int32(p.Results.tile_width);
                desc.tile_info.tile_height = int32(p.Results.tile_height);
            end
            desc.has_compression = int32(0);

            if ~isempty(p.Results.camera)
                c = p.Results.camera;
                camstruct = c.to_struct();
                sink = ptiff_sink_create_camera(patht, desc, camstruct);
                delete_ptiff_camera(camstruct);
            else
                sink = ptiff_sink_create(patht, desc);
            end
            if isempty(sink)
                delete_ptiff_image_descriptor(desc);
                error('ptiff:ImageCreate', 'ptiff_sink_create(%s) failed', patht);
            end

            img = Image();
            img.path = patht;
            img.sink_ = sink;
            img.desc_ = desc;
            if ~isempty(p.Results.camera)
                img.camera_ = p.Results.camera;
            end
        end
    end

    methods
        function width = width(obj)
            obj = obj.ensure_desc_();
            width = obj.desc_.width;
        end

        function height = height(obj)
            obj = obj.ensure_desc_();
            height = obj.desc_.height;
        end

        function ch = channel_count(obj)
            obj = obj.ensure_desc_();
            ch = obj.desc_.channel_count;
        end

        function pt = pixel_type(obj)
            obj = obj.ensure_desc_();
            pt = obj.desc_.pixel_type;
        end

        function c = camera(obj)
            c = obj.camera_;
        end

        function n = tile_columns(obj)
            if ~isempty(obj.source_)
                n = ptiff_source_tile_columns(obj.source_);
            elseif ~isempty(obj.sink_)
                n = ptiff_sink_tile_columns(obj.sink_);
            else
                n = 0;
            end
        end

        function n = tile_rows(obj)
            if ~isempty(obj.source_)
                n = ptiff_source_tile_rows(obj.source_);
            elseif ~isempty(obj.sink_)
                n = ptiff_sink_tile_rows(obj.sink_);
            else
                n = 0;
            end
        end

        function n = tile_byte_size(obj)
            if ~isempty(obj.source_)
                n = ptiff_source_tile_byte_size(obj.source_);
            elseif ~isempty(obj.sink_)
                n = ptiff_sink_tile_byte_size(obj.sink_);
            else
                n = 0;
            end
        end

        function t = read_tile(obj, column, row)
            %READ_TILE  Read the tile at grid position (column, row) -> Tile.
            if isempty(obj.source_)
                error('ptiff:ReadTile', 'image not open for reading: %s', obj.path);
            end
            bs = obj.tile_byte_size();
            [rc, data, ~] = ptiff_source_read_tile(obj.source_, int32(column), int32(row), uint8(zeros(1, bs)));
            if rc ~= 0
                error('ptiff:ReadTile', 'ptiff_source_read_tile(%d,%d) failed (rc=%d)', column, row, rc);
            end
            t = Tile(column, row, data, bs);
        end

        function rc = write_tile(obj, column, row, data)
            %WRITE_TILE  Write data to the tile at grid position (column, row).
            if isempty(obj.sink_)
                error('ptiff:WriteTile', 'image not open for writing: %s', obj.path);
            end
            rc = ptiff_sink_write_tile(obj.sink_, int32(column), int32(row), uint8(data));
            if rc ~= 0
                error('ptiff:WriteTile', 'ptiff_sink_write_tile(%d,%d) failed (rc=%d)', column, row, rc);
            end
        end

        function close(obj)
            %CLOSE  Release the underlying source/sink handles (idempotent).
            if ~isempty(obj.source_)
                try
                    ptiff_source_close(obj.source_);
                catch
                end
                obj.source_ = [];
            end
            if ~isempty(obj.sink_)
                try
                    ptiff_sink_close(obj.sink_);
                catch
                end
                obj.sink_ = [];
                if ~isempty(obj.desc_)
                    delete_ptiff_image_descriptor(obj.desc_);
                    obj.desc_ = [];
                end
            end
        end
    end

    methods (Access = private)
        function obj = ensure_desc_(obj)
            if isempty(obj.desc_) && ~isempty(obj.source_)
                d = new_ptiff_image_descriptor();
                ptiff_source_descriptor(obj.source_, d);
                obj.desc_ = d;
            end
        end

        function obj = load_descriptor_(obj)
            obj = obj.ensure_desc_();
        end
    end
end

File: bindings/go/ptiff/image.go
package ptiff

import "fmt"

// Image is a PTIFF image: metadata plus tile access.
//
// It mirrors Python's `ptiff.Image` over the SWIG low-level source/sink
// surface. Use OpenImage to read a TIFF/BigTIFF and CreateImage to write a new
// one. Reading exposes the structured camera calibration via Camera and the raw
// tile bytes via ReadTile.
type Image struct {
	path   string
	src    Struct_SS_ptiff_source // non-nil when open for reading
	sink   Struct_SS_ptiff_sink   // non-nil when open for writing
	camera Camera
	desc   Ptiff_image_descriptor
}

// OpenImage opens the TIFF/BigTIFF file at path for reading and returns an
// Image wrapping the SWIG source handle. On C-ABI failure a non-zero error code
// is returned as an error.
func OpenImage(path string) (*Image, error) {
	var errOut int
	src := Ptiff_source_open(path, &errOut)
	if src.Swigcptr() == 0 {
		return nil, fmt.Errorf("ptiff_source_open: error code %d", errOut)
	}

	cam, _ := OpenPathCamera(path) // best-effort; a missing camera is not fatal

	desc := NewPtiff_image_descriptor()
	if rc := Ptiff_source_descriptor(src, desc); rc != 0 {
		Ptiff_source_close(src)
		return nil, fmt.Errorf("ptiff_source_descriptor: error code %d", rc)
	}

	return &Image{path: path, src: src, camera: cam, desc: desc}, nil
}

// CreateOptions carries optional settings for CreateImage (tiling and camera).
type CreateOptions struct {
	TileWidth    uint
	TileHeight   uint
	ChannelCount uint
	Camera       *Camera // when set, the calibration is persisted alongside
}

// CreateImage creates a new image at path with the given size and pixel type.
// A zero-value CreateOptions produces an untiled, single-channel image. On
// C-ABI failure a non-zero error code is returned as an error.
func CreateImage(path string, width, height uint, pixelType PixelType, opts CreateOptions) (*Image, error) {
	channels := opts.ChannelCount
	if channels == 0 {
		channels = 1
	}
	desc := NewPtiff_image_descriptor()
	desc.SetWidth(width)
	desc.SetHeight(height)
	desc.SetPixel_type(int(pixelType))
	desc.SetChannel_count(channels)
	if opts.TileWidth > 0 && opts.TileHeight > 0 {
		tiles := NewPtiff_tile_info()
		tiles.SetTile_width(opts.TileWidth)
		tiles.SetTile_height(opts.TileHeight)
		desc.SetHas_tile_info(1)
		desc.SetTile_info(tiles)
	}
	desc.SetHas_compression(0)
	desc.SetCompression(0)

	var sink Struct_SS_ptiff_sink
	if opts.Camera != nil {
		cam := opts.Camera.toMetadata()
		sink = Ptiff_sink_create_camera(path, desc, cam)
	} else {
		sink = Ptiff_sink_create(path, desc)
	}
	if sink.Swigcptr() == 0 {
		return nil, fmt.Errorf("ptiff_sink_create failed for %q", path)
	}

	var cam Camera
	if opts.Camera != nil {
		cam = *opts.Camera
	}
	return &Image{path: path, sink: sink, camera: cam, desc: desc}, nil
}

// Path returns the file path this Image points at.
func (img *Image) Path() string { return img.path }

// Width returns the image width in pixels.
func (img *Image) Width() uint { return img.desc.GetWidth() }

// Height returns the image height in pixels.
func (img *Image) Height() uint { return img.desc.GetHeight() }

// ChannelCount returns the number of channels.
func (img *Image) ChannelCount() uint { return img.desc.GetChannel_count() }

// PixelType returns the pixel type (a PTIFF_PIXEL_* value).
func (img *Image) PixelType() PixelType { return PixelType(img.desc.GetPixel_type()) }

// Camera returns the structured camera calibration. For an opened image it is
// read from the file's `ptiff.camera.*` extension fields; for a created image
// it is whatever CreateOptions.Camera carried (possibly zero-valued).
func (img *Image) Camera() Camera { return img.camera }

// TileColumns returns the number of tile columns (0 when untiled).
func (img *Image) TileColumns() uint {
	if img.src != nil {
		return Ptiff_source_tile_columns(img.src)
	}
	if img.sink != nil {
		return Ptiff_sink_tile_columns(img.sink)
	}
	return 0
}

// TileRows returns the number of tile rows (0 when untiled).
func (img *Image) TileRows() uint {
	if img.src != nil {
		return Ptiff_source_tile_rows(img.src)
	}
	if img.sink != nil {
		return Ptiff_sink_tile_rows(img.sink)
	}
	return 0
}

// TileByteSize returns the byte-payload size of a single tile (0 when untiled).
func (img *Image) TileByteSize() int64 {
	if img.src != nil {
		return Ptiff_source_tile_byte_size(img.src)
	}
	if img.sink != nil {
		return Ptiff_sink_tile_byte_size(img.sink)
	}
	return 0
}

// ReadTile reads the tile at grid position (column, row) and returns its raw
// bytes. It requires the image to be open for reading (OpenImage).
func (img *Image) ReadTile(column, row uint) (Tile, error) {
	if img.src == nil {
		return Tile{}, fmt.Errorf("image not open for reading: %q", img.path)
	}
	n := img.TileByteSize()
	buf := make([]byte, int(n))
	var bytesRead int64
	if rc := Ptiff_source_read_tile(img.src, column, row, buf, &bytesRead); rc != 0 {
		return Tile{}, fmt.Errorf("ptiff_source_read_tile(%d,%d): error code %d", column, row, rc)
	}
	return NewTile(column, row, buf, bytesRead), nil
}

// WriteTile writes data to the tile at grid position (column, row). It requires
// the image to be open for writing (CreateImage).
func (img *Image) WriteTile(column, row uint, data []byte) error {
	if img.sink == nil {
		return fmt.Errorf("image not open for writing: %q", img.path)
	}
	if rc := Ptiff_sink_write_tile(img.sink, column, row, data); rc != 0 {
		return fmt.Errorf("ptiff_sink_write_tile(%d,%d): error code %d", column, row, rc)
	}
	return nil
}

// Close releases the underlying source/sink handles. It is safe to call
// multiple times; closed handles are released and zeroed.
func (img *Image) Close() {
	if img.src != nil {
		Ptiff_source_close(img.src)
		img.src = nil
	}
	if img.sink != nil {
		Ptiff_sink_close(img.sink)
		img.sink = nil
	}
}
